### DISCUSSION TITLE
Reports and Tags -> AttributeError: _TaggableManager

### DISCUSSION THREAD

**Discussion by martinum4** [netbox-community/netbox#6444] [Q&A] [answered]
Hello,

does anyone know how i can use tag-data in reports?
Can i "dry run" a report with python directly when i try to debug it?

Background: 
We have integrated devices from a service provider in our Netbox-instance. These devices have a tag "foreign ownership" since we own some devices of the same type too, our devices are untagged.
For own own devices Asset-Tag and Serial Number is required, while for foreign devices only the Serial Number is relevant.
Thus i wanted to create a report that first filters if the device is foreign owned.

I tried the following:
```
import re # because i want to check asset tag formatting in the future

from dcim.choices import DeviceStatusChoices
from dcim.models import Device
from extras.reports import Report
class InventoryDataReport(Report):
        description = "<german text>"
        def test_device_inventorydata(self):
                for device in Device.objects.filter(status=DeviceStatusChoices.STATUS_ACTIVE):
                        if str(device.device_role.name) != ("Infrastruktur"): # we don't want passive infrastructure 
                                if int(device.tags.id) != "3":
…<other code>…
```

If i try to run this report i get the following error:
```
An exception occurred: AttributeError: '_TaggableManager' object has no attribute 'id' Traceback (most recent call last): File "/opt/netbox-2.11.1/netbox/extras/reports.py", line 232, in run test_method() File "/opt/netbox-2.11.1/netbox/reports/InventoryData.py", line 19, in test_device_inventorydata if int(device.tags.id) != "3": AttributeError: '_TaggableManager' object has no attribute 'id'
```

So, how can i query for the tags correctly? Is it even possible?

Kind regards

Martin

**Comment by candlerb** [comment, marked as answer]
> Can i "dry run" a report with python directly when i try to debug it?

You can [run a report from the command line](https://netbox.readthedocs.io/en/stable/additional-features/reports/#via-the-cli).  You can also use [nbshell](https://netbox.readthedocs.io/en/stable/administration/netbox-shell/) to interact with the python models interactively, for finding out how these things work.

I don't know what you mean by "dry run" in this context - reports should not be making any changes to the database.

> So, how can i query for the tags correctly? Is it even possible?

Yes, of course it's possible: Netbox does it :-)

```
>>> d = Device.objects.get(id=1)
>>> d
<Device: nuc1>
>>> d.tags
<taggit.managers._TaggableManager object at 0x7f967691c470>
>>> d.tags.all()
<RestrictedQuerySet [<Tag: prom_node>]>
>>> d.tags.all()[0]
<Tag: prom_node>
>>> d.tags.all()[0].id
7
>>> d.tags.all()[0].name
'prom_node'
>>> [t.id for t in d.tags.all()]
[7]
>>> help(Tag)
... <scrollable help on the Tag model>
q

>>> dir(d.tags)
... <snip long list of methods>
>>> d.tags.names
<bound method _TaggableManager.names of <taggit.managers._TaggableManager object at 0x7f967691c4a8>>
>>> d.tags.names()
<RestrictedQuerySet ['prom_node']>
>>> list(d.tags.names())
['prom_node']
```

Or in a Django ORM query:

```
>>> Device.objects.filter(tags__name__in=['prom_node']).count()
3
>>> Device.objects.filter(tags__name='prom_node').count()
3
>>> Device.objects.filter(tags__id=7).count()
3
```

There are examples in the [reports](https://github.com/netbox-community/reports) repository, e.g. [here](https://github.com/netbox-community/reports/blob/master/reports/virtualization-reports/vm_counts.py#L42).

**Reply by martinum4** [reply]
Thanks a lot, your answer helped me immensely, the netbox shell was what i was looking for, however i had to be in the virtual environment mentioned in the install files.

For anyone having the same problem as me:
```
if str(device.device_role.name) != "Infrastruktur":
  if ("Fremdeigentum" in list(device.tags.names())) and str(device.serial) is "":
    self.log_warning(device, "Fremdeigentum ohne SN/Foreign ownership devices without serial number")
```
